compiler+runtime/src/cpp/jank/read/lex.cpp

```diff
@@ -1205,20 +1205,33 @@ namespace jank::read::lex
           {
             ++pos;
             auto const after_dd_colon(peek());
-            if(after_dd_colon.is_ok() && after_dd_colon.expect_ok().character == '/')
+
+            /* Invalid ":: " source found. */
+            if(after_dd_colon.is_err())
+            {
+              return error::lex_invalid_keyword(
+                "A auto-resolved keyword must contain a valid symbol after the '::'.",
+                { token_start, pos });
+            }
+
+            auto const ch{ after_dd_colon.expect_ok().character };
+
+            /* Invalid ::/, ::), etc. pattern found. */
+            if(after_dd_colon.is_ok() && (ch == '/' || !is_symbol_char(ch)))
             {
-              /* Invalid ::/ pattern found. Consume the entire invalid token. */
+              /* Consume the entire invalid token. */
               while(true)
               {
                 auto const result(convert_to_codepoint(file.substr(pos), pos));
-                if(result.is_err() || !is_symbol_char(result.expect_ok().character))
+                if(result.is_err() || !is_symbol_char(ch))
                 {
                   break;
                 }
                 pos += result.expect_ok().len;
               }
-              return error::lex_invalid_keyword("An auto-resolved keyword may not start with '/'.",
-                                                { token_start, pos });
+              return error::lex_invalid_keyword(
+                util::format("An auto-resolved keyword may not start with '{}'.", ch),
```

Comment:
We have the information here for a better error. We know that we have a `::` and that we have one of two cases:

1. `::/`, so we can provide an error saying the namespace cannot be empty for the keyword.
2. `::` followed by a non-symbol char, so we can provide an error saying there's an unexpected end to that keyword. This is an incomplete keyword.

I think trying to word it as what a keyword can't start with is less helpful to the user. The user isn't going to type `::)` because they think a keyword can start with that. They're going to do that because they forgot to provide the rest of the keyword.

Reply:
Yes they can. Auto-resolved keyword namespaces match namespace aliases (that sentence is a tongue twister).

For example:
```clojure
(require '[jank.compiler :as c])
::c/foo ; :jank.compiler/foo
```